File: app/src/main/java/com/example/creditcalculator/PatternLockView.java
package com.example.creditcalculator;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;

import java.util.ArrayList;
import java.util.List;

/**
 * Lightweight 3x3 application pattern lock.
 * The saved value is only the node sequence; AppPreferences hashes it before storage.
 */
public class PatternLockView extends View {
    public interface OnPatternCompleteListener {
        void onPatternComplete(String pattern);
    }

    private final Paint dotPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint selectedPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint linePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final float[] xs = new float[9];
    private final float[] ys = new float[9];
    private final List<Integer> selected = new ArrayList<>();
    private OnPatternCompleteListener listener;
    private float dotRadius;
    private float hitRadius;
    private float currentX;
    private float currentY;
    private boolean drawing;
    private boolean error;

    public PatternLockView(Context context) {
        super(context);
        init();
    }

    public PatternLockView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public PatternLockView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        setFocusable(true);
        setClickable(true);
        dotPaint.setStyle(Paint.Style.STROKE);
        dotPaint.setStrokeWidth(dp(2));
        dotPaint.setColor(ContextCompat.getColor(getContext(), R.color.border));

        selectedPaint.setStyle(Paint.Style.FILL);
        selectedPaint.setColor(ContextCompat.getColor(getContext(), R.color.primary));

        linePaint.setStyle(Paint.Style.STROKE);
        linePaint.setStrokeCap(Paint.Cap.ROUND);
        linePaint.setStrokeJoin(Paint.Join.ROUND);
        linePaint.setStrokeWidth(dp(5));
        linePaint.setColor(ContextCompat.getColor(getContext(), R.color.primary));
        linePaint.setAlpha(205);

        setMinimumHeight(dp(250));
    }

    public void setOnPatternCompleteListener(OnPatternCompleteListener listener) {
        this.listener = listener;
    }

    public void reset() {
        selected.clear();
        drawing = false;
        error = false;
        invalidate();
    }

    public void showError() {
        error = true;
        drawing = false;
        invalidate();
        postDelayed(this::reset, 700);
    }

    public String getPatternString() {
        StringBuilder out = new StringBuilder();
        for (int index : selected) out.append(index + 1);
        return out.toString();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        float usableW = Math.max(1, w - getPaddingLeft() - getPaddingRight());
        float usableH = Math.max(1, h - getPaddingTop() - getPaddingBottom());
        float size = Math.min(usableW, usableH);
        float left = getPaddingLeft() + (usableW - size) / 2f;
        float top = getPaddingTop() + (usableH - size) / 2f;
        float step = size / 3f;
        dotRadius = Math.max(dp(8), Math.min(dp(15), step * 0.11f));
        hitRadius = Math.max(dp(30), step * 0.34f);
        for (int row = 0; row < 3; row++) {
            for (int col = 0; col < 3; col++) {
                int i = row * 3 + col;
                xs[i] = left + step * (col + 0.5f);
                ys[i] = top + step * (row + 0.5f);
            }
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        int activeColor = ContextCompat.getColor(getContext(), error ? R.color.danger : R.color.primary);
        selectedPaint.setColor(activeColor);
        linePaint.setColor(activeColor);

        if (!selected.isEmpty()) {
            Path path = new Path();
            int first = selected.get(0);
            path.moveTo(xs[first], ys[first]);
            for (int i = 1; i < selected.size(); i++) {
                int node = selected.get(i);
                path.lineTo(xs[node], ys[node]);
            }
            if (drawing) path.lineTo(currentX, currentY);
            canvas.drawPath(path, linePaint);
        }

        for (int i = 0; i < 9; i++) {
            boolean active = selected.contains(i);
            if (active) {
                canvas.drawCircle(xs[i], ys[i], dotRadius * 1.35f, selectedPaint);
                Paint inner = new Paint(Paint.ANTI_ALIAS_FLAG);
                inner.setStyle(Paint.Style.FILL);
                inner.setColor(ContextCompat.getColor(getContext(), R.color.card_background));
                canvas.drawCircle(xs[i], ys[i], dotRadius * 0.52f, inner);
            } else {
                canvas.drawCircle(xs[i], ys[i], dotRadius, dotPaint);
            }
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (!isEnabled()) return false;
        currentX = event.getX();
        currentY = event.getY();
        switch (event.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                if (!selected.isEmpty()) reset();
                drawing = true;
                addHitNode(currentX, currentY);
                invalidate();
                return true;
            case MotionEvent.ACTION_MOVE:
                addHitNode(currentX, currentY);
                invalidate();
                return true;
            case MotionEvent.ACTION_UP:
                drawing = false;
                addHitNode(currentX, currentY);
                invalidate();
                performClick();
                if (listener != null) listener.onPatternComplete(getPatternString());
                return true;
            case MotionEvent.ACTION_CANCEL:
                drawing = false;
                invalidate();
                return true;
            default:
                return super.onTouchEvent(event);
        }
    }

    @Override
    public boolean performClick() {
        super.performClick();
        return true;
    }

    private void addHitNode(float x, float y) {
        int hit = hitTest(x, y);
        if (hit < 0 || selected.contains(hit)) return;
        if (!selected.isEmpty()) {
            int last = selected.get(selected.size() - 1);
            int middle = middleNode(last, hit);
            if (middle >= 0 && !selected.contains(middle)) selected.add(middle);
        }
        selected.add(hit);
    }

    private int hitTest(float x, float y) {
        float best = Float.MAX_VALUE;
        int result = -1;
        for (int i = 0; i < 9; i++) {
            float dx = x - xs[i];
            float dy = y - ys[i];
            float d2 = dx * dx + dy * dy;
            if (d2 <= hitRadius * hitRadius && d2 < best) {
                best = d2;
                result = i;
            }
        }
        return result;
    }

    /** Adds the skipped centre node for straight standard Android-style moves. */
    private int middleNode(int a, int b) {
        int ar = a / 3, ac = a % 3;
        int br = b / 3, bc = b % 3;
        int dr = br - ar, dc = bc - ac;
        if (Math.abs(dr) == 2 && dc == 0) return ((ar + br) / 2) * 3 + ac;
        if (Math.abs(dc) == 2 && dr == 0) return ar * 3 + ((ac + bc) / 2);
        if (Math.abs(dr) == 2 && Math.abs(dc) == 2) return ((ar + br) / 2) * 3 + ((ac + bc) / 2);
        return -1;
    }

    private float dp(int value) {
        return value * getResources().getDisplayMetrics().density;
    }
}

File: MathProgress/app/src/main/java/com/mathprogress/app/PatternLockView.kt
package com.mathprogress.app

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.view.MotionEvent
import android.view.View
import kotlin.math.hypot

class PatternLockView(
    context: Context,
    private val darkMode: Boolean,
    private val onComplete: (String) -> Unit
) : View(context) {
    private val selected = mutableListOf<Int>()
    private val points = mutableListOf<Pair<Float,Float>>()
    private var currentX = 0f
    private var currentY = 0f
    private var drawing = false
    private var error = false
    private val line = Paint(Paint.ANTI_ALIAS_FLAG).apply { strokeWidth = dp(5f); strokeCap = Paint.Cap.ROUND }
    private val dot = Paint(Paint.ANTI_ALIAS_FLAG)

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val w = MeasureSpec.getSize(widthMeasureSpec)
        setMeasuredDimension(w, w.coerceAtMost(dp(330f).toInt()))
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        buildPoints()
        line.color = if (error) Color.rgb(215,63,70) else Color.rgb(99,91,255)
        if (selected.size > 1) for (i in 0 until selected.lastIndex) {
            val a=points[selected[i]]; val b=points[selected[i+1]]; canvas.drawLine(a.first,a.second,b.first,b.second,line)
        }
        if (drawing && selected.isNotEmpty()) {
            val a=points[selected.last()]; canvas.drawLine(a.first,a.second,currentX,currentY,line)
        }
        for (i in points.indices) {
            val (x,y)=points[i]
            dot.style=Paint.Style.FILL
            dot.color = if (i in selected) line.color else if (darkMode) Color.rgb(66,68,78) else Color.rgb(222,224,232)
            canvas.drawCircle(x,y,dp(if(i in selected)13f else 10f),dot)
            if (i !in selected) {
                dot.style=Paint.Style.STROKE; dot.strokeWidth=dp(2f); dot.color=if(darkMode)Color.rgb(112,115,128) else Color.rgb(170,173,184)
                canvas.drawCircle(x,y,dp(18f),dot)
            }
        }
    }

    override fun onTouchEvent(e: MotionEvent): Boolean {
        when(e.action) {
            MotionEvent.ACTION_DOWN -> { selected.clear(); error=false; drawing=true; currentX=e.x;currentY=e.y;addHit(e.x,e.y);invalidate();return true }
            MotionEvent.ACTION_MOVE -> { currentX=e.x;currentY=e.y;addHit(e.x,e.y);invalidate();return true }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                drawing=false; currentX=e.x;currentY=e.y; invalidate()
                if(selected.size>=4) onComplete(selected.joinToString("-")) else { error=true; invalidate(); postDelayed({reset()},650) }
                return true
            }
        }
        return true
    }

    fun showError() { error=true; invalidate(); postDelayed({reset()},650) }
    fun reset() { selected.clear();error=false;drawing=false;invalidate() }

    private fun buildPoints(){
        if(points.isNotEmpty())return
        val usableW=width.toFloat(); val usableH=height.toFloat(); val xs=listOf(usableW*.2f,usableW*.5f,usableW*.8f); val ys=listOf(usableH*.18f,usableH*.5f,usableH*.82f)
        for(y in ys)for(x in xs)points+=x to y
    }
    private fun addHit(x:Float,y:Float){
        buildPoints(); val idx=points.indices.firstOrNull{ i->i !in selected && hypot(x-points[i].first,y-points[i].second)<dp(30f)}
        if(idx!=null)selected+=idx
    }
    override fun performClick(): Boolean { super.performClick(); return true }
    private fun dp(v:Float)=v*resources.displayMetrics.density
}
